File: 자료구조 부수기/BOJ_2493.java
package BOJ_2493;

import java.io.*;
import java.util.ArrayDeque;
import java.util.Iterator;
import java.util.StringTokenizer;

public class BOJ_2493 {

    // https://www.acmicpc.net/problem/2493
    // input
    private static BufferedReader br;

    // variables
    private static int N;
    private static int[] arr;

    private static class Top {
        int num;
        int idx;

        private Top(int num, int idx) {
            this.num = num;
            this.idx = idx;
        }
    } // End of Top class

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\알고리즘\\JavaAlgorithm\\src\\BOJ_2493\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        ArrayDeque<Top> que = new ArrayDeque<>();
        for (int i = 0; i < N; i++) {
            while (!que.isEmpty() && que.peekLast().num <= arr[i]) {
                que.removeLast();
            }

            if (que.isEmpty()) {
                sb.append(0);
            } else {

                Iterator<Top> iter = que.descendingIterator();
                while (iter.hasNext()) {
                    Top next = iter.next();
                    if (next.num >= arr[i]) {
                        sb.append(next.idx);
                        break;
                    }
                }
            }

            sb.append(' ');
            que.offerLast(new Top(arr[i], i + 1));
        }

        return sb.toString();
    } // End of solve()

    private static void input() throws IOException {
        N = Integer.parseInt(br.readLine());
        arr = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
    } // End of input()
} // End of Main class

File: 자료구조 부수기/BOJ 2493.kt
package BOJ_2493

import java.io.File
import java.util.StringTokenizer


// input
private var br = System.`in`.bufferedReader()

// variables
private var N = 0
private lateinit var arr: IntArray

private data class Top(var num: Int, var idx: Int)

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_2493\\res.txt"
    br = File(path).bufferedReader()
    val bw = System.out.bufferedWriter()

    input()

    bw.write(solve())
    bw.close()
} // End of main()

private fun solve(): String {
    val sb = StringBuilder()

    // 앞에 있는 탑 보다, 새로들어오는 탑의 높이가 더 클 경우 앞은 제거.
    val que = ArrayDeque<Top>()
    que.add(Top(arr[0], 1))
    sb.append(0).append(' ')

    for (i in 1 until N) {

        while (que.isNotEmpty() && que.last().num <= arr[i]) {
            que.removeLast()
        }

        if (que.isEmpty()) {
            sb.append(0)
        } else {
            val size = que.size
            for (j in size - 1 downTo 0) {
                if (que.get(j).num >= arr[i]) {
                    sb.append(que.get(j).idx)
                    break
                }
            }
        }

        que.addLast(Top(arr[i], i + 1))
        sb.append(' ')
    }

    return sb.toString()
} // End of solve()

private fun input() {
    N = br.readLine().toInt()
    arr = IntArray(N)

    StringTokenizer(br.readLine()).let {
        for (i in 0 until N) {
            arr[i] = it.nextToken().toInt()
        }
    }
} // End of input()
